Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -884,47 +997,109 @@ public async Task<object> EvaluateDynamicProperty(dynamic schemaObj, Type knownT
         /// </summary>
-        /// <param name="treeNodeKey">The TreeNode's key that tree walker is currently walking.</param>
-        private Task CommitCurrentTreeNode(string treeNodeKey)
+        /// <param name="treeNodeKey">The TreeNode's key that tree walker is about to visit.</param>
+        /// <param name="treeNode">The TreeNode that tree walker is about to visit.</param>
+        private async Task CommitCurrentTreeNode(string treeNodeKey, TreeNode treeNode)
         {
-            // TODO: Consider adding a "nodeStep" to save if we already completed the BeforeVisitNode step and should skip straight to VisitNode.
-            return this.Parameters.ForgeState.Set<string>(CurrentTreeNodeSuffix, treeNodeKey);
+            List<KeyValuePair<string, object>> itemsToPersist = new List<KeyValuePair<string, object>>();
+
+            // Handle revisit/cycle case if this node has been previously completed and we aren't rehydrating.
+            if (treeNode.Actions != null && this.hasSessionRehydrated)
+            {
+                // Check if all actions already have an action response.
+                foreach (KeyValuePair<string, TreeAction> kvp in treeNode.Actions)
+                {
+                    string treeActionKey = kvp.Key;
+                    TreeAction treeAction = kvp.Value;
+                    ActionResponse actionResponse = await this.GetOutputAsync(treeActionKey).ConfigureAwait(false);
+
+                    if (actionResponse == null)
+                    {
+                        // If this session has rehydrated and actionResponse is null, then this must be the first time visiting this TreeNode.
+                        itemsToPersist.Clear();
+                        break;
+                    }
+
+                    // If this session has rehydrated and actionResponses exist, then this TreeNode was previously visited.
+                    // Clear the ActionResponse and Intermediates, and persist PreviousActionResponse.
+                    itemsToPersist.Add(new KeyValuePair<string, object>(treeActionKey + ActionResponseSuffix, null));
+                    itemsToPersist.Add(new KeyValuePair<string, object>(treeActionKey + IntermediatesSuffix, null));
+                    itemsToPersist.Add(new KeyValuePair<string, object>(treeActionKey + PreviousActionResponseSuffix, actionResponse));
+                }
+            }
+
+            itemsToPersist.Add(new KeyValuePair<string, object>(CurrentTreeNodeSuffix, treeNodeKey));
+
+            await this.Parameters.ForgeState.SetRange(itemsToPersist);
+            this.hasSessionRehydrated = true;
         }
 
         /// <summary>
-        /// Commits the last tree action to the forgeState.
+        /// Gets the previously committed TreeInput if it exists, otherwise persists the incoming TreeInput.
```

Comment:
>Gets the previously committed TreeInput if it exists, otherwise persists the incoming TreeInput. [](start = 12, length = 96)

If the previously committed TreeInput is different than the incoming TreeInput, say some parameter value changes in the new schema. Shall we use the new one, which has the most recently requirement?

Reply:
No, we use the previous one intentionally. Otherwise there would be no need to persist the TreeInput.
Here is the reasoning for this decision. For example, you pass in a TreeInput, and you rely on the value of the TreeInput to walk to NodeA, and then failover.
Upon rehydration, let's say the TreeInput is calculated differently (perhaps indicating we should have walked to NodeB instead).
Because we are already at NodeA, updating the TreeInput could break our assumptions.
We want avoid that, so we persist the TreeInput and use that instead.

---
In reply to: [356762006](https://github.com/microsoft/Forge/pull/6#discussion_r356762006) [](ancestors = 356762006)